### ISSUE TITLE
Submodule add command fails

### ISSUE THREAD

**Issue by MythreyaK** [git-for-windows/git#2448]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.24.1.windows.2
cpu: x86_64
built from commit: 992f0773022527b1b0cb1e0c13aec97dd5248053
sizeof-long: 4
sizeof-size_t: 8
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.18363.535]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
Editor Option: VisualStudioCode
Custom Editor Path: 
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: WinSSL
CRLF Option: LFOnly
Bash Terminal Option: ConHost
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Enabled
Enable Builtin Interactive Add: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?
Nothing I can think of. 

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other
    Powershell

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
$ git init .
$ git submodule add <insert repo> 
```
 - What did you expect to occur after running these commands?
    Git clones the repo and adds a submodule

 - What actually happened instead?

```
C:/Program Files/Git/mingw64/libexec/git-core\git-submodule: line 1121: cmd_: command not found
```

**Comment by dscho** [comment]
I cannot reproduce this. Can you try with https://github.com/git-for-windows/git/releases/tag/v2.25.0-rc0.windows.1?

**Comment by MythreyaK** [comment]
> I cannot reproduce this. 

That's surprising, my issue persisted over a reinstall! 

Either way, issue persists. `C:/Program Files/Git/mingw64/libexec/git-core\git-submodule: line 1171: cmd_: command not found`

Just in case, I switched off windows defender off too (to prevent force-ASLR).

**Comment by dscho** [comment]
> I switched off windows defender off too (to prevent force-ASLR).

That is most definitely an interesting thing about your environment that might be related
to the issue you're seeing.

To be able to run Unix shell scripts, we use MSYS2's Bash. This means that Unix shell scripting relies on the MSYS2 runtime, which is a friendly fork of the Cygwin runtime. And the matter of the fact is that Cygwin's/MSYS2's runtime won't be able to `fork()` new processes if ASLR is force-enabled. See e.g. [this illuminating email thread on the Cygwin mailing list](http://cygwin.1069669.n5.nabble.com/ASLR-breaks-cygwin-td99276.html).

In other words, this forced ASLR business is most likely what bites you, and I doubt that it really is turned off if you still experience the issue.

The most realistic work-around is to force-turn-off ASLR in your setup.

If that is not possible, you have two options:

- move away from submodules, or
- work on turning the [`git-submodule.sh` script](https://github.com/git/git/blob/v2.24.1/git-submodule.sh) into a built-in written in C.

Should you choose the latter, I would be willing to mentor you.

**Comment by MythreyaK** [comment]
Yes, I force-switched-off ASLR system wide, and it worked! I must've missed adding some binary to the exception list. 

As for the second option, it sounds very interesting! I'm not sure what level of expertise, prior knowledge and experience is required. I'm fairly new to tackling large codebases and have had little experience in making one. I am interested in taking it up, if it's not too far too big a bite for me to handle. I am up for the learning experience and the challenge though! Thank you very much for your kind offer!

**Comment by dscho** [comment]
> I'm not sure what level of expertise, prior knowledge and experience is required.

Since I offered to mentor, only a little knowledge of C and shell scripting is required.

It may very well be a large project, time-wise, though... :-)

So how to start with the project? Let's first look into the story arc, so to say, as converting a shell script into a built-in is never a single patch.

The most important thing to realize (at least in my mind) is that it is not so much a conversion as a re-implementation, and the best way to go about it is incremental.

Take for example the `git submodule init` subcommand. It has been re-implemented in C a long time ago, and the shell script version merely parses the command-line arguments and then passes them on to the built-in `git submodule--helper` command:

https://github.com/git/git/blob/v2.24.1/git-submodule.sh#L361-L390

It probably would make sense to study the initial re-implementation (or "port"): 3604242f080a813d6f20a7394def422d1e55b30e

From what I see, `cmd_add`, `cmd_init`, `cmd_deinit`, (partially) `cmd_update`, (partially) `cmd_set_branch` and `cmd_summary` still need to be ported while `cmd_foreach`, `cmd_status`, `cmd_sync` and `cmd_absorbgitdirs` already use the `submodule--helper`-provided built-in code.

The easiest of the sub-commands that still need to be ported might be `cmd_set_branch`. I see in the git.git fork of the contributor who used to work actively on the porting effort (before switching jobs, away from Git) that there are [some branches](https://github.com/stefanbeller/git/branches) that might or might not have useful commits in them, but given that `set_branch` already calls the `submodule--helper` a couple of times and does little on its own, it might make sense to just do that port "from scratch".

The idea here will be to pass the options using the construct `${GIT_QUIET:+--quiet}` (which expands to `--quiet` if `$GIT_QUIET` would expand to a non-empty string) to the `submodule--helper`, and to add a new sub-command to said helper, by implementing `static int set_branch(int argc, const char **argv, const char *prefix)` in `builtin/submodule--helper.c` and appending the corresponding entry to [the `commands` array](https://github.com/git/git/blob/v2.24.1/builtin/submodule--helper.c#L2217-L2239).

Likewise, you should be able to figure out which functions are used by the `submodule--helper` to perform the actions asked for by the current shell script function `cmd_set_branch`, and call them directly from the newly-implemented `set_branch()` function in `builtin/submodule--helper.c`.

Finally, I would suggest to open a PR as soon as you have anything to show, even if it is not working yet. The hardest part, as with every project, is to get started, once you have something to build on, it is very easy to ask for help (and I encourage you to do precisely that).

**Comment by dscho** [comment]
As the root cause has been identified, and as we cannot really "fix" the MSYS2 runtime (see #1412 for more on this), I'll close this ticket.

Of course, nothing prevents us from talking about porting more of `git submodule` to C, even though it is closed.

**Comment by MythreyaK** [comment]
I had come across this ticket before! I set some exceptions to let git clone work, and never thought about that again. That's the reason I mentioned ASLR here, assuming this was probably the cause. So this issue cannot be solved by a port to C? I was under the impression that porting to C would prevent a …. wait... Bash forks! That's why it can't be fixed here. Well, that's a bummer!

I've done some digging, and while I haven't been able to grasp the work to be done yet, thanks to your insight, I now understand where to begin! In the meantime, I forked this repo and managed to get a builds working on Azure pipelines. How do I test any code I write, since the build on pipelines takes over 7 hours, all combined?

Also, if it's okay, I have an exam coming up on the 12th, so I won't be able to dedicate good productive chunks of time writing code. Again, thank you very much for mentoring me, this is a really great opportunity!

**Comment by dscho** [comment]
> So this issue cannot be solved by a port to C?

Not really, as Git assumes that it has access to a Unix shell, and Git for Windows relies on MSYS2's Bash for that. And MSYS2's Bash relies on the MSYS2 runtime which requires ASLR to be turned off...

> Bash forks! That's why it can't be fixed here. Well, that's a bummer!

Exactly.

> In the meantime, I forked this repo and managed to get a builds working on Azure pipelines. How do I test any code I write, since the build on pipelines takes over 7 hours, all combined?

Great! To test, I would suggest to focus on the scripts in the test suite that are related to submodules (`echo t/t[0-9]*submodule*.sh`).

The pipeline should not take over 7 hours, as it is parallelized. It typically completes after 35-40 minutes over here. Is that not the case for you?

> Also, if it's okay, I have an exam coming up on the 12th, so I won't be able to dedicate good productive chunks of time writing code. Again, thank you very much for mentoring me, this is a really great opportunity!

Take your time. I will be delighted with whatever progress you make, whenever it is convenient for you.
